R/tables.R

```diff
@@ -1,117 +1,274 @@
+#'@importFrom cli cat_line
+#'@importFrom crayon cyan magenta
+#'@importFrom tibble as.tibble
+#'@export
+print.eidith_tbl <- function(x,...){
+  note_cols <- attributes(x)$notes
+  print(as.tibble(unclass(x)))
+  if(length(note_cols > 0)){
+    cat_line("")
+    cat_line(cyan("There are notes attached to this dataframe which may contain important information! Please look at the following column(s):"))
+    cat_line(paste("    ", magenta(names(x)[note_cols])))
+  }
+  duplicate_keys <- attributes(x)$duplicate_keys
+  duplicate_indices <- attributes(x)$duplicate_indices
+  if(length(duplicate_keys > 0)){
+    cat_line("")
+    cat_line(red("IMPORTANT: There are multiple rows with duplicate unique ID's in this table!"))
+    cat_line(black("              The duplicate IDs are:"))
+    cat_line(paste("                  ", red(duplicate_keys)))
+    cat_line(black("              This affects the following rows:"))
+    cat_line(paste("                  ", red(duplicate_indices)))
+  }
+}
+
+
+
+
+
+
+
+fix_classes <- function(table) {
+  logical_cols <- names(table)[names(table) %in% logical_vars]
+  date_cols <-  names(table)[names(table) %in% date_vars]
+  datetime_cols <- names(table)[names(table)  %in% datetime_vars]
+  if(length(logical_cols) != 0) table <- mutate_at(table, vars(logical_cols), funs_("as.logical"))
+  if(length(date_cols) != 0) table <- mutate_at(table, vars(date_cols), funs(quicktime))
+  if(length(datetime_cols) != 0) table <- mutate_at(table, vars(datetime_cols), funs(quicktime2))
+  return(table)
+}
+
+#' @export
+#' @rdname ed_table
+ed_table <- function(table, ...) {
+  ed_table_(table, .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+#' @rdname ed_table
+ed_events <- function(...) {
+  ed_table_("events", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+#' @rdname ed_table
+ed_animals <- function(...) {
+  ed_table_("animals", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+#' @rdname ed_table
+ed_specimens <- function(...) {
+  ed_table_("specimens", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+#' @rdname ed_table
+ed_tests <- function(...) {
+  ed_table_("tests", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+#' @rdname ed_table
+ed_viruses <- function(...) {
+  ed_table_("viruses", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+#' @rdname ed_table
+ed_testspecimen <- function(...) {
+  ed_table_("test_specimen_ids", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+#' @importFrom DBI dbReadTable
+ed2_key_errors <- function() {
+  dbReadTable(eidith_db()$con, "p2_unique_id_errors")
+}
+
+#' @export
+ed2_events <- function(...){
+  ed_table_("events_2", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+ed2_animals <- function(...) {
+  ed_table_("animals_2", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+ed2_specimens <- function(...) {
+  ed_table_("specimens_2", .dots = lazyeval::lazy_dots(...))
+}
+
+#' @export
+ed2_animal_production <- function(...){

```

Comment:
This would be a good table to have a demo of how to gather/tidy the data in our recipes.